File: user/src/main/java/io/philo/domain/service/UserService.java
package io.philo.domain.service;

import io.philo.domain.entity.User;
import io.philo.domain.repository.UserRepository;
import io.philo.shop.error.UnauthorizedException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class UserService {

  private final UserRepository repository;

  public long createUser(String email, String name, String address, String password) {
    var user = new User(email, name, address, password);
    repository.save(user);
    return user.getId();
  }

  public void login(String inputEmail, String inputPassword) {
    User user = repository.findByEmailOrThrow(inputEmail);
    validateCredential(inputEmail, user);
  }

  private static void validateCredential(String inputEmail, User user) {
    if (!isRightCredential(inputEmail, user)) {
      throw new UnauthorizedException("유효한 로그인 정보가 아닙니다.");
    }
  }

  private static boolean isRightCredential(String inputEmail, User user) {
    if (user.getEmail().equals(inputEmail)) {
      if (user.getEncodedPassword().equals(inputEmail)) {
        return true;
      }
      return false;
    }
    return false;
  }
}

File: user/src/main/kotlin/io/philo/domain/service/UserService.kt
package io.philo.domain.service

import io.philo.domain.entity.User
import io.philo.domain.repository.UserRepository
import io.philo.shop.error.EntityNotFoundException
import io.philo.shop.error.UnauthorizedException
import io.philo.shop.user.dto.UserPassportResponse
import io.philo.support.JwtManager
import org.springframework.stereotype.Service

@Service
class UserService(private val jwtManager: JwtManager,
                  private val repository: UserRepository) {


    fun createUser(
        email: String,
        name: String,
        address: String,
        password: String
    ): Long {

        val user = User(email, name, address, password)

        repository.save(user)

        return user.id!!
    }

    fun login(inputEmail: String, inputPassword: String): String {

        val user = repository.findByEmail(inputEmail) ?: throw EntityNotFoundException(inputEmail)

        validateCredential(inputPassword, user)

        // todo 더 암호화하기... 이렇게 되면 보안에 취약하다
        val subject = user.id.toString()
        val newAccessToken = jwtManager.createAccessToken(subject)

        return newAccessToken
    }

    fun isValidToken(accessToken: String): Boolean {
        return jwtManager.isValidToken(accessToken)
    }

    fun passport(accessToken: String): UserPassportResponse {

        return if (jwtManager.isValidToken(accessToken)) {
            val userInfo = jwtManager.parse(accessToken)
            val user: User? = repository.findById(userInfo.toLong()).orElseGet(null)
            if (user == null) {
                UserPassportResponse.OfInvalid()
            } else {
                UserPassportResponse.OfValid(user.id!!, user.name, user.email)
            }
        } else {
            UserPassportResponse.OfInvalid()
        }
    }

    private fun validateCredential(inputPassword: String, user: User) {
        if (isCorrectCredential(inputPassword, user).not()) {
            throw UnauthorizedException("유효한 로그인 정보가 아닙니다.")
        }
    }

    private fun isCorrectCredential(inputPassword: String, user: User): Boolean {
        return user.isSamePassword(inputPassword)
    }
}
